servant/src/Servant/API/Modifiers.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeOperators       #-}
+module Servant.API.Modifiers (
+    -- * Required / optional argument
+    Required, Optional,
+    FoldRequired, FoldRequired',
+    -- * Lenient / strict parsing
+    Lenient, Strict,
+    FoldLenient, FoldLenient',
+    -- * Utilities
+    RequiredArgument,
+    foldRequiredArgument,
+    unfoldRequiredArgument,
+    RequestArgument,
+    unfoldRequestArgument,
+    ) where
+
+import Data.Proxy           (Proxy (..))
+import Data.Singletons.Bool (SBool (..), SBoolI (..))
+import Data.Text            (Text)
+import Data.Type.Bool       (If)
+
+-- | Required argument. Not wrapped.
+data Required
+
+-- | Optional argument. Wrapped in 'Maybe'.
+data Optional
+
+-- | Fold modifier list to decide whether argument is required.
+--
+-- >>> :kind! FoldRequired '[Required, Description "something"]
+-- FoldRequired '[Required, Description "something"] :: Bool
+-- = 'True
+--
+-- >>> :kind! FoldRequired '[Required, Optional]
+-- FoldRequired '[Required, Optional] :: Bool
+-- = 'False
+--
+-- >>> :kind! FoldRequired '[]
+-- FoldRequired '[] :: Bool
+-- = 'False
+--
+type FoldRequired mods = FoldRequired' 'False mods
```

Comment:
Hmm, so here, another option is to have a map:

``` [IsRequired := 'True, Description := "something"] ```

(And optionally, `type Required = IsRequired := 'True ; type Optional = IsRequired := 'False`)

Which simplifies the implementation here, insofar as instead of having separate `FoldX`s, you can just have a generic "Lookup".

Reply:
Using map-like is an option.

TL;DR, current approach:
- pros: there isn't ill-kinded situations like `IsRequired := "true"`
- cons: slight boilerplate in the implementation (not use!)

IMHO pros outweight cons here.

---

I personally dislike "dynamic" feeling of `(:=) :: * -> k -> *`. Nothing really prevents you from saying `'[IsRequired := "true"]`. OTOH, having *typed* labels is a great improvements over `(:=) :: Symbol -> k -> *`, proposed earlier, as then typos in label name would get unnoticed.

Think about term level, we could have two APIs:

```haskell
defaultOptions &
  & required .~ True
  & description ?~ "Foo"
-- or
required <> description "Foo"
```

The type-level list approach is like the `Monoid` one.
Having all "type-safety" of lens one (value type is dependent on the key!) is tricky.
My gut feeling that in GHC `>= 8.0` one could encode that, but IMHO it's not worth that.